Allow usage with repositories using release branches

weierophinney opened the pull request:
This patch adds a script at the start of the entrypoint that detects whether or not we should build docs for the given push, using the following rules:

- If the branch is "master", BUILD.
- If the branch name does not match `\d+\.\d+\.x`, we DO NOT BUILD.
- If the latest tag (per semver) matches the major/minor version of the branch, WE BUILD.
- Otherwise, we DO NOT BUILD.

The script is fairly primitive; I didn't want to bring in any extra deps (e.g., via Composer), but is based on work I've done locally to try out approaches. I'm willing to add assertions as necessary; mainly wanting some initial feedback.

michalbundyra commented on github-actions/docs/determine_if_we_need_to_build.php:
```diff
@@ -0,0 +1,82 @@
+$compareRefNames = function ($a, $b) {
+    $a = preg_replace('/\.x$/', '.0', $a);
+    $b = preg_replace('/\.x$/', '.0', $b);
+    return $a <=> $b;
```
not sure if this is right here, see for example:

echo '1.10.0' <=> '1.9.0'; // -1 (we expect 1)
echo '1.9.0' <=> '1.8.0'; // 1 (correct)

weierophinney replied:
You're right - I meant to use `version_compare` here! Thanks!

michalbundyra commented on github-actions/docs/determine_if_we_need_to_build.php:
```diff
@@ -0,0 +1,82 @@
+
+if (! preg_match('/^\d+\.\d+\.x$/', $ref)) {
+    // Not a release branch
+    skipDocs(); // redundant; placed here to note that skipDocs exits
```
I believe the additional comment should be on line with `exit(0)`  (as above), the same for couple more places below

michalbundyra commented on github-actions/docs/determine_if_we_need_to_build.php:
```diff
@@ -0,0 +1,82 @@
+    return (bool) preg_match('/^\d+\.\d+\.x$/', $branch);
+});
+$mostRecent = array_pop($branches);
+if ($mostRecent === $ref) {
```
I might be missing something here, but I am not sure if we should build docs on the latest branch. For example in laminas/automatic-releases we have 1.3.x and latest release is 1.2.4, so branch 1.3.x. is "like old develop" and it could contains some unreleased changes (new features, etc..)

It's quite confusing atm as I am not sure where we should really push the documentation fixes.
And we cannot really build on two different branches, as documentation can be different on these.

From above I can see that we build on the latest, or... previous if it is the latest release.

froschdesign replied:
> It's quite confusing atm as I am not sure where we should really push the documentation fixes.

At the moment we need no new release to update the documentation. It's done on the _default branch_ which was the branch with the name "master" so far. In the new branching model the default branch is the "next release branch".
Therefore the question is correct here: where to push the corrections for the documentation?

weierophinney replied:
@froschdesign Pull requests, by default, go to the default branch. With the new branching model, that branch is version.next, instead of version.current. With the old model, this made submitting documentation changes easy, as you could just submit against the default branch, and then we could merge immediately — the onus was on the person merging to then also remember to merge against the develop branch.

Under the new model, there are two possibilities:

- The author of the patch submits against the stable branch. It is then up to the maintainer to ensure that the patch also goes up to the default branch (as normally, merge-ups are only created during a release).

- The author of the patch submits against the default branch. It is then up to the maintainer to decide if the patch should also be merged to the current stable branch for immediate publication.

The second is easier logistically, as it ensures that the patch will be available for all versions going forwards, and it leaves the decision about whether this is a substantive change to the maintainer, versus the author of the documentation patch.

weierophinney commented:
Pushed changes that reflect the feedback. In particular:

- Now using version_compare for sorting, instead of spaceship operator.
- Fixed an issue with how the comparison function is declared and consumed (was using a closure previously, but not passing it to the function consuming it).
- Removed the logic to fetch branches, and now instead always check the release branch against the latest tag (per semver). This simplifies the main logic of the script tremendously.

weierophinney commented on github-actions/docs/README.md:
```diff
@@ -79,7 +79,7 @@ name: docs-build
   push:
     branches:
-      - master
+      - '*'
```
So, with branches, you don't have as much leeway. I've modified it to be '*.x', but that could still trap some invalid branch names (e.g., "foo-bar.x"); the code in the PHP script will catch any remaining such branches. I've also explicitly added "master" back.

I've now also added a "paths" section, and indicated each of:

- 'docs/book/**'
- 'mkdocs.yml'
- 'doc/book/**' (for some legacy structures we still have)

This should definitely help limit the number of builds triggered; thanks for the links!

weierophinney commented:
Closing in favor of 50, which is a cleaner solution.
